File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.cpp
#include <string>
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> solution(vector<vector<int>> edges) { 

    // int max = max_element(edges.begin(), edges.end()) - edges.begin();
    // int temp = edges[max][0] > edges[max][1] ? 0 : 1;
    int max = 0;
    for(int i=0;i<edges.size();i++){
        if(edges[i][0] > max) max = edges[i][0];
        if(edges[i][1] > max) max = edges[i][1];
    }
    
    vector<int> count(max+1,0);
    vector<int> take(max+1,0);
    vector<int> out(max+1,0);

   for(int i=0;i<edges.size();i++){
        count[edges[i][1]]++;
        count[edges[i][0]]--;
       out[edges[i][0]]++;
        take[edges[i][1]]++;  
    }

    // int add = min_element(count.begin(), count.end()) - count.begin();
    int add = 0;
    for(int i = 1; i < count.size(); i++) {
        if(count[add] > count[i]) add = i;
    }

    int k=0;
    for(int i=0;i<edges.size();i++){
        if(edges[i][0]==add){
            k++;
        }
    }

    int stick=0; 
    int eight=0; 
    int donut=0; 

    //cout<<k;

    for(int i=1;i<count.size();i++){
        if(count[i] == 1 && take[i] == 1){
            stick++;
            // cout<<i<<'\n';
        }
    }
     /*for(int i=0;i<count.size();i++){
       cout<<count[i]<<'\t'<<take[i]<<'\n';
    }*/

    for(int i=0;i<count.size();i++){
        if(count[i]>0){
            count[i]--;
            take[i]--;
        }
    }

    for(int i=0;i<count.size();i++){
        if(count[i]==0&&take[i]==2){
            eight++;
        } 
        else if(count[i]==1&&take[i]==1){
            stick++;
        }
    }
    donut=k-eight-stick;

     /*for(int i=0;i<count.size();i++){
       cout<<count[i]<<'\t'<<take[i]<<'\n';
    }*/

    vector<int> answer;
    answer={add,donut,stick,eight};
    return answer;
}

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.java
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    public int[] solution(int[][] edges) {
        int[] answer;
        HashMap<Integer, ArrayList<Integer>> pointOut = new HashMap<>();
        HashMap<Integer, int[]> count = new HashMap<>();
        
        for(int i = 0; i < edges.length; i++) {
            pointOut.putIfAbsent(edges[i][0], new ArrayList<Integer>());
            pointOut.get(edges[i][0]).add(edges[i][1]);
            
            count.putIfAbsent(edges[i][0], new int[]{0, 0});
            count.putIfAbsent(edges[i][1], new int[]{0, 0});
            count.get(edges[i][0])[1]++;
            count.get(edges[i][1])[0]++;
        }
        
        int point = 0;
        int donut = 0;
        int stick = 0;
        int eight = 0;
        int totalCount = 0;
        
        for(Integer i : count.keySet()) {
            int[] temp = count.get(i);
            if(temp[0] == 0 && temp[1] > 1) {
                point = i;
                totalCount = pointOut.get(i).size();
            } else {
                if(temp[0] >= 2 && temp[1] >= 2) eight++;
                if(temp[1] == 0) stick++;
            }
        }
        
        donut = totalCount - stick - eight;
        
        answer = new int[]{point, donut, stick, eight};
        
        return answer;
    }
}